File: dbus-tiny-client.cpp
#include <dbus-tiny.h>

#include <string>
#include <iostream>
#include <boost/format.hpp>
#include <boost/program_options.hpp>

int main(int argc, const char **argv)
{
	try
	{
		boost::program_options::options_description	options("usage");
		boost::program_options::positional_options_description positional_options;

		try
		{
			bool introspect = false;
			std::string service;
			std::string interface;
			std::string rv1;
			std::string string_call_void;
			std::string string_call_string;
			std::string call_x_1;
			std::string call_x_2;
			std::string signal_string;
			std::vector<std::string> arguments;

			options.add_options()
				("service,s",				boost::program_options::value<std::string>(&service),					"service to use")
				("interface,i",				boost::program_options::value<std::string>(&interface),					"interface to use")
				("introspect,I",			boost::program_options::bool_switch(&introspect)->implicit_value(true),	"show introspection")
				("string-call-void,v",		boost::program_options::value<std::string>(&string_call_void),			"call method taking no arguments returning string")
				("string-call-string,c",	boost::program_options::value<std::string>(&string_call_string),		"call method taking string returning string")
				("call-x-1,1",				boost::program_options::value<std::string>(&call_x_1),					"call method taking u32,u32,string,string returning u64,u32,u32,string,double")
				("call-x-2,2",				boost::program_options::value<std::string>(&call_x_2),					"call method taking void returning u64,3xstring,4xdouble")
				("signal-string,S",			boost::program_options::value<std::string>(&signal_string),				"send signal with string parameter")
				("argument",				boost::program_options::value<std::vector<std::string>>(&arguments),	"specify method arguments");

			positional_options.add("argument", -1);

			boost::program_options::variables_map varmap;
			boost::program_options::store(boost::program_options::command_line_parser(argc, argv).options(options).positional(positional_options).run(), varmap);
			boost::program_options::notify(varmap);

			DbusTinyClient dbus_client;

			if(service.length() == 0)
				service = "/org/freedesktop/DBus/dummy";

			if(introspect)
			{
				dbus_client.send_void(service, "org.freedesktop.DBus.Introspectable", "Introspect");
				rv1 = dbus_client.receive_string();
				std::cerr << rv1 << std::endl;
			}
			else if(string_call_void.length() > 0)
			{
				dbus_client.send_void(service, interface, string_call_void);
				rv1 = dbus_client.receive_string();
				std::cerr << rv1 << std::endl;
			}
			else if(string_call_string.length() > 0)
			{
				if(arguments.size() != 1)
					throw("call-string-string needs one argument");

				dbus_client.send_string(service, interface, string_call_string, arguments.at(0));
				rv1 = dbus_client.receive_string();
				std::cerr << rv1 << std::endl;
			}
			else if(call_x_1.length() > 0)
			{
				uint32_t p0, p1;
				uint64_t r0;
				uint32_t r1, r2;
				std::string r3;
				double r4;

				if(arguments.size() != 4)
					throw("call-x-1 needs 4 arguments");

				try
				{
					p0 = std::stoi(arguments.at(0));
					p1 = std::stoi(arguments.at(1));
				}
				catch(const std::invalid_argument &)
				{
					throw("invalid numeric argument");
				}
				catch(const std::out_of_range &)
				{
					throw("invalid numeric argument");
				}

				dbus_client.send_uint32_uint32_string_string(service, interface, call_x_1, p0, p1, arguments.at(2), arguments.at(3));
				dbus_client.receive_uint64_uint32_uint32_string_double(r0, r1, r2, r3, r4);

				std::cerr << boost::format("%llu / %lu / %lu / %s / %f\n") % r0 % r1 % r2 % r3 % r4;
			}
			else if(call_x_2.length() > 0)
			{
				uint64_t r0;
				std::string r1, r2, r3;
				double r4, r5, r6, r7;

				if(arguments.size() != 0)
					throw("call-x-2 needs no arguments");

				dbus_client.send_void(service, interface, call_x_2);
				dbus_client.receive_uint64_x3string_x4double(r0, r1, r2, r3, r4, r5, r6, r7);

				std::cerr << boost::format("%llu / %s / %s / %s / %f / %f / %f / %f\n") % r0 % r1 % r2 % r3 % r4 % r5 % r6 % r7;
			}
			else if(signal_string.length() > 0)
			{
				if(arguments.size() != 1)
					throw("signal-string needs one argument");

				dbus_client.signal_string(service, interface, signal_string, arguments.at(0));
			}
		}
		catch(const boost::program_options::error &e)
		{
			throw((boost::format("program option exception: %s\n%s") % e.what() % options).str());
		}
		catch(const DbusTinyException &e)
		{
			throw((boost::format("error: %s") % e.what()).str());
		}
		catch(const std::exception &e)
		{
			throw((boost::format("standard exception: %s") % e.what()).str());
		}
		catch(const std::string &e)
		{
			throw((boost::format("error: %s ") % e).str());
		}
		catch(const char *e)
		{
			throw((boost::format("error: %s") % e).str());
		}
		catch(...)
		{
			throw(std::string("unknown exception"));
		}
	}
	catch(const std::string &e)
	{
		std::cerr << "dbus-tiny-client: " << e << std::endl;
	}
}

File: dbus-tiny-client.pl
#!/usr/bin/perl -w

use strict;
use warnings;

use Getopt::Long qw(:config gnu_compat noignore_case);
use Try::Tiny;
use Scalar::Util;
use DBUS::Tiny;

use Data::Dumper;

my($introspect);
my($service);
my($interface);
my($string_call_void);
my($string_call_string);
my($call_x_1);
my($call_x_2);
my($signal_string);
my(@arguments);
my($argument);
my($rv1);

GetOptions(
		"service|s=s"				=> \$service,
		"interface|i=s"				=> \$interface,
		"introspect|I"				=> \$introspect,
		"string-call-void|v=s"		=> \$string_call_void,
		"string-call-string|c=s"	=> \$string_call_string,
		"call-x-1|1=s"				=> \$call_x_1,
		"call-x-2|2=s"				=> \$call_x_2,
		"signal-string|S=s"			=> \$signal_string,
		"argument|a=s"				=> \@arguments,
);

try
{
	my($dbus_client) = new DBUS::Tiny::DbusTinyClient;

	if(!defined($service) && !defined($signal_string))
	{
		print STDERR ("usage: dbus-tiny-client.pl <options> <arguments>\n");
		print STDERR ("       -s|--service <service>\n");
		print STDERR ("       -i|--interface <interface> (optional for method calling)\n");
		print STDERR ("       -I|--introspect (introspect)\n");
		print STDERR ("       -v|--string-call-void <method>\n");
		print STDERR ("       -c|--string-call-string <method>\n");
		print STDERR ("       -1|--call-x-1 <method>\n");
		print STDERR ("       -2|--call-x-2 <method>\n");
		print STDERR ("       -S|--signal-string <method>\n");
		print STDERR ("       -a|--argument <argument> (may be repeated)\n");
	}

	$service = "/org/freedesktop/DBus/dummy" if(!defined($service));
	$interface = "" if(!defined($interface));

	while(defined($argument = shift(@ARGV)))
	{
		push(@arguments, $argument);
	}

	if(defined($introspect))
	{
		$dbus_client->send_void($service, "org.freedesktop.DBus.Introspectable", "Introspect");
		$rv1 = $dbus_client->receive_string();
		printf ("output: %s\n", $rv1);
	}
	elsif(defined($string_call_void))
	{
		$dbus_client->send_void($service, $interface, $string_call_void);
		$rv1 = $dbus_client->receive_string();
		printf ("output: %s\n", $rv1);
	}
	elsif(defined($string_call_string))
	{
		die("string-call-string takes one argument: string") if(scalar(@arguments) != 1);

		$dbus_client->send_string($service, $interface, $string_call_string, $arguments[0]);
		$rv1 = $dbus_client->receive_string();
		printf ("output: %s\n", $rv1);
	}
	elsif(defined($call_x_1))
	{
		my($r1, $r2, $r3, $r4, $r5);

		die("call_x_1 takes four arguments: int, int, string, string") if(scalar(@arguments) != 4);
		die("call_x_1 argument one should be an integer number") if(!Scalar::Util::looks_like_number($arguments[0]));
		die("call_x_1 argument two should be an integer number") if(!Scalar::Util::looks_like_number($arguments[1]));

		$dbus_client->send_uint32_uint32_string_string($service, $interface, $call_x_1, int($arguments[0]), int($arguments[1]), $arguments[2], $arguments[3]);
		$dbus_client->receive_uint64_uint32_uint32_string_double();

		$r1 = $dbus_client->get_rv_uint64_0();
		$r2 = $dbus_client->get_rv_uint32_0();
		$r3 = $dbus_client->get_rv_uint32_1();
		$r5 = $dbus_client->get_rv_string_0();
		$r4 = $dbus_client->get_rv_double_0();

		printf ("results: %llu/%lu/%lu/%f:%s\n", $r1, $r2, $r3, $r4, $r5);
	}
	elsif(defined($call_x_2))
	{
		my($r0, $r1, $r2, $r3, $r4, $r5, $r6, $r7);

		die("call_x_2 takes no arguments") if(scalar(@arguments) != 0);

		$dbus_client->send_void($service, $interface, $call_x_2);
		$dbus_client->receive_uint64_x3string_x4double();

		$r0 = $dbus_client->get_rv_uint64_0();
		$r1 = $dbus_client->get_rv_string_0();
		$r2 = $dbus_client->get_rv_string_1();
		$r3 = $dbus_client->get_rv_string_2();
		$r4 = $dbus_client->get_rv_double_0();
		$r5 = $dbus_client->get_rv_double_1();
		$r6 = $dbus_client->get_rv_double_2();
		$r7 = $dbus_client->get_rv_double_3();

		printf ("%llu / %s / %s / %s / %f / %f / %f / %f\n", $r0, $r1, $r2, $r3, $r4, $r5, $r6, $r7);
	}
	elsif(defined($signal_string))
	{
		die("signal_string takes one argument: string") if(scalar(@arguments) != 1);

		$dbus_client->signal_string($service, $interface, $signal_string, $arguments[0]);
	}
}
catch
{
	printf STDERR ("failed: %s\n", $_);
	exit(1);
};
